Update protocols, make drivers follow protocols

Bastian-Krause opened the pull request:
**Description**
Make method args of `CommandProtocol`, `FileTransferProtocol` consistent/reflect driver reality. Now let drivers implementing protocols follow them strictly.

This PR includes converting `run()`'s/`run_check()`'s `codec`/`decodeerrors` kwargs to driver attributes.

**Checklist**
- [x] The arguments and description in doc/configuration.rst have been updated
- [ ] PR has been tested (parts of it)

Bastian-Krause commented:
Force-push:
- mentioned `codec` and `decodeerrors` in CHANGES.rst's "Breaking" section
- make `SSHDriver.run()` pass timeout argument, set default timeout to 30 as specified by CommandProtocol

jluebbe commented on labgrid/driver/sshdriver.py:
```diff
@@ -103,24 +105,24 @@ def _check_master(self):
-    def run(self, cmd, codec="utf-8", decodeerrors="strict", timeout=None): # pylint: disable=unused-argument
-        return self._run(cmd, codec=codec, decodeerrors=decodeerrors)
+    @step(args=['command'], result=True)
+    def run(self, command: str, *, timeout: int = 30):
```
Why do we want to have a default timeout for the SSHDriver? Compared to the serial based drivers, SSH easily detects a crashed target, so introducing such a short timeout might be surprising.

Bastian-Krause replied:
I thought it would be a good thing to have consistent timeouts over all drivers implementing the `CommandProtocol`. I understand now that timeouts are a way to detect a crashed target in case of serial communication. In that case I will remove the default timeout from the `CommandProtocol` and set it to None for the `SSHDriver`, alright?

Bastian-Krause replied:
Implemented as suggested, see below.

Bastian-Krause commented:
Force-push:
- rebased on current master
- remove default timeout of `CommandProtocol` `run()` and `run_check()`
- set default timeout of `SSHDriver` `run()` and `run_check()` to `None`
- moved `decodeerrors` to `CommandMixin`

Note: the new process handling API being discussed in #552 will return bytes instead of strings. There is no need to decode for (sub)string search or regex matching when this is used.

Bastian-Krause commented:
Force-push: rebased on master

Emantor commented:
Rebased on current master to fix merge conflict in `sshdriver.py`.

Emantor commented:
@Bastian-Krause please recheck the rebase was done correctly.

Bastian-Krause commented:
@Emantor Rebase looks good.

Rebased it once again on current master.

Bastian-Krause commented:
Ping?

jluebbe commented on CHANGES.rst:
```diff
@@ -56,6 +56,9 @@ Breaking changes in 0.3.0
   renamed to `USBStorageDriver`.
   A deprecated `NetworkUSBStorageDriver` exists temporarily for compatibility
   reasons.
+- ``codec`` and ``decodeerrors`` are no longer ``run()``/``run_check()``
+  kwargs, but driver (``codec``) and `CommandMixin` (``decodeerrors``)
+  attributes.
```
Why is codec a driver attribute and decodeerrors on CommandMixin?

This should also describe how a user should update their code and configuration. Is the encoding changeable at runtime?

jluebbe commented on doc/configuration.rst:
```diff
@@ -817,6 +817,7 @@ Arguments:
     to pass before sending a newline to device.
   - console_ready (regex): optional, pattern used by the kernel to inform
     the user that a console can be activated by pressing enter.
+  - codec (str, default="utf-8"): optional, encoding used to decode output bytes
```
Is "output" the bytes sent or received by labgrid? Shouldn't the encoding apply to both directions?

jluebbe commented on labgrid/driver/commandmixin.py:
```diff
@@ -13,6 +13,7 @@ class CommandMixin:
 
     def __attrs_post_init__(self):
         super().__attrs_post_init__()
+        self.decodeerrors = 'strict'
```
This is not an attrs attribute, so it can't be specified during construction. What is the reason for the difference?
